it/src/test/java/com/google/cloud/teleport/it/spanner/DefaultSpannerResourceManagerTest.java

```diff
@@ -0,0 +1,395 @@
+        testManager.readTableRecords("Singers", "SingerId", "FirstName", "LastName");
+
+    // assert
+    ImmutableList<Struct> expected = ImmutableList.of(struct1, struct2);
+    assertThat(actual).containsExactlyElementsIn(expected);
+  }
+
+  @Test
+  public void testReadRecordsShouldThrowExceptionWhenSpannerReadFails()
+      throws ExecutionException, InterruptedException {
+    // arrange
+    prepareTable();
+    when(spanner.getDatabaseClient(any()).singleUse().read(any(), any(), any()))
+        .thenThrow(SpannerException.class);
+
+    // act & assert
+    assertThrows(
+        SpannerResourceManagerException.class,
+        () -> testManager.readTableRecords("Singers", "SingerId", "FirstName", "LastName"));
+  }
+
+  @Test
+  public void testCleanupAllShouldThrowExceptionWhenSpannerDeleteInstanceFails() {
+    // arrange
+    doThrow(SpannerException.class).when(instanceAdminClient).deleteInstance(any());
+    when(spanner.getInstanceAdminClient()).thenReturn(instanceAdminClient);
+    testManager = new DefaultSpannerResourceManager(spanner, TEST_ID, PROJECT_ID, REGION);
+
+    // act & assert
+    assertThrows(SpannerResourceManagerException.class, () -> testManager.cleanupAll());
+  }
+
+  @Test
+  public void testCleanupAllShouldWorkWhenSpannerDeleteInstanceSucceeds() {
+    // arrange
+    doNothing().when(instanceAdminClient).deleteInstance(any());
+    when(spanner.getInstanceAdminClient()).thenReturn(instanceAdminClient);
+    testManager = new DefaultSpannerResourceManager(spanner, TEST_ID, PROJECT_ID, REGION);
+
+    // act
+    testManager.cleanupAll();
+
+    // assert
+    verify(spanner.getInstanceAdminClient()).deleteInstance(any());
+    verify(spanner).close();
+  }
+
+  @Test
+  public void testManagerShouldBeUnusableAfterCleanup() {
+    // arrange
+    doNothing().when(instanceAdminClient).deleteInstance(any());
+    when(spanner.getInstanceAdminClient()).thenReturn(instanceAdminClient);
+    when(spanner.isClosed()).thenReturn(true);
+    testManager = new DefaultSpannerResourceManager(spanner, TEST_ID, PROJECT_ID, REGION);
+    testManager.cleanupAll();
+    String statement =
+        "CREATE TABLE Singers (\n"
+            + "  SingerId   INT64 NOT NULL,\n"
+            + "  FirstName  STRING(1024),\n"
+            + "  LastName   STRING(1024),\n"
+            + ") PRIMARY KEY (SingerId)";
+    Mutation testMutation =
+        Mutation.newInsertOrUpdateBuilder("SingerId")
+            .set("SingerId")
+            .to(1)
+            .set("FirstName")
+            .to("Marc")
+            .set("LastName")
+            .to("Richards")
+            .build();
+
+    // act & assert
+    assertThrows(IllegalStateException.class, () -> testManager.createTable(statement));
+    assertThrows(
+        IllegalStateException.class, () -> testManager.readTableRecords("Singers", "SingerId"));
+    assertThrows(IllegalStateException.class, () -> testManager.write(testMutation));
+  }
+
```

Comment:
Maybe I missed it, but could you also add tests for when a user tries to read or write before calling `createTable`?

Reply:
Could you add it for all the read/write methods, not just one of each? 

I know that these methods call the others, but tests shouldn't be written as if we know exactly how each method is implemented. In theory, each method could have its own unique implementation without calling the other.

Reply:
There still needs to be one for `readTableRecords(String, Iterable<String>)`. It looks like there's only one for `readTableRecords(String, String...)`. Feel free to just add a similar assertion to the same test method.